File: Web Services and Middleware Technologies/Middleware/Pyro3Server/service/service.py
import json

import Pyro.core

from domain.Grade import Grade
from domain.Student import Student
from domain.Subject import Subject


class Service(Pyro.core.ObjBase):

    def __init__(self, student_repo, subject_repo, grade_repo):
        Pyro.core.ObjBase.__init__(self)
        self.__student_repo = student_repo
        self.__subject_repo = subject_repo
        self.__grade_repo = grade_repo

    # @expose
    def add_student(self, first_name, last_name, email):
        student = Student(first_name=first_name, last_name=last_name, email=email)
        self.__student_repo.add(student)

    # @expose
    def delete_student(self, student_id):
        self.__student_repo.delete(student_id)

    # @expose
    def update_student(self, student_id, first_name, last_name, email):
        self.__student_repo.update(student_id, first_name, last_name, email)

    # @expose
    def get_all_students(self):
        jsons = {'data': list(map(lambda s: s.to_json(), self.__student_repo.get_all()))}
        return json.dumps(jsons)

    # @expose
    def add_subject(self, code, name):
        subject = Subject(subject_id=code, name=name)
        self.__subject_repo.add(subject)

    # @expose
    def delete_subject(self, subject_id):
        self.__subject_repo.delete(subject_id)

    # @expose
    def update_subject(self, subject_id, name):
        self.__subject_repo.update(subject_id, name)

    # @expose
    def get_all_subjects(self):
        jsons = {'data': list(map(lambda s: s.to_json(), self.__subject_repo.get_all()))}
        return json.dumps(jsons)

    # @expose
    def add_grade(self, student_id, subject_id, value):
        grade = Grade(student_id=student_id, subject_id=subject_id, value=value)
        self.__grade_repo.add(grade)

    # @expose
    def update_grade(self, grade_id, value):
        self.__grade_repo.update(grade_id, value)

    # @expose
    def get_all_grades(self):
        jsons = {'data': list(map(lambda g: g.to_json(), self.__grade_repo.get_all()))}
        return json.dumps(jsons)

File: Web Services and Middleware Technologies/Web Services/CSharpServer/CSharpServer/service/Service.cs
using System;
using CookComputing.XmlRpc;
using CSharpServer.domain;
using CSharpServer.repository;
using CSharpServer.util;

namespace CSharpServer.service
{
    public class Service : XmlRpcListenerService
    {
        private readonly StudentRepository _studentRepository;
        private readonly SubjectRepository _subjectRepository;
        private readonly GradeRepository _gradeRepository;

        public Service(StudentRepository studentRepository, SubjectRepository subjectRepository,
            GradeRepository gradeRepository)
        {
            this._studentRepository = studentRepository;
            this._subjectRepository = subjectRepository;
            this._gradeRepository = gradeRepository;
        }

        [XmlRpcMethod("service.AddStudent", Description = "Adds a student with given attributes to the database")]
        public void AddStudent(string firstName, string lastName, string email)
        {
            Console.WriteLine("New AddStudent Request " + DateTime.Now);
            var student = new Student {FirstName = firstName, LastName = lastName, Email = email};
            this._studentRepository.Add(student);
        }

        [XmlRpcMethod("service.DeleteStudent", Description = "Deletes student with given id from the database")]
        public void DeleteStudent(int studentId)
        {
            Console.WriteLine("New DeleteStudent Request " + DateTime.Now);
            this._studentRepository.Delete(studentId);
        }

        [XmlRpcMethod("service.UpdateStudent", Description = "Updates student with given id")]
        public void UpdateStudent(int studentId, string firstName, string lastName, string email)
        {
            Console.WriteLine("New UpdateStudent Request " + DateTime.Now);
            this._studentRepository.Update(studentId, firstName, lastName, email);
        }

        [XmlRpcMethod("service.GetAllStudents", Description = "Retrieves all students from the database")]
        public XmlRpcStruct[] GetAllStudents()
        {
            Console.WriteLine("New GetAllStudents Request " + DateTime.Now);
            return Converter.StudentsToXmlRpcStructs(this._studentRepository.GetAll());
        }

        [XmlRpcMethod("service.AddSubject", Description = "Adds a subject with given attributes to the database")]
        public void AddSubject(string code, string name)
        {
            Console.WriteLine("New AddSubject Request " + DateTime.Now);
            var subject = new Subject {SubjectId = code, Name = name};
            this._subjectRepository.Add(subject);
        }

        [XmlRpcMethod("service.DeleteSubject", Description = "Deletes subject with given id from the database")]
        public void DeleteSubject(string subjectId)
        {
            Console.WriteLine("New DeleteSubject Request " + DateTime.Now);
            this._subjectRepository.Delete(subjectId);
        }

        [XmlRpcMethod("service.UpdateSubject", Description = "Updates subject with given id")]
        public void UpdateSubject(string subjectId, string name)
        {
            Console.WriteLine("New UpdateSubject Request " + DateTime.Now);
            this._subjectRepository.Update(subjectId, name);
        }

        [XmlRpcMethod("service.GetAllSubjects", Description = "Retrieves all subjects from the database")]
        public XmlRpcStruct[] GetAllSubjects()
        {
            Console.WriteLine("New GetAllSubjects Request " + DateTime.Now);
            return Converter.SubjectsToXmlRpcStructs(this._subjectRepository.GetAll());
        }

        [XmlRpcMethod("service.AddGrade", Description = "Adds a grade with given attributes to the database")]
        public void AddGrade(int studentId, string subjectId, double value)
        {
            Console.WriteLine("New AddGrade Request " + DateTime.Now);
            var grade = new Grade {StudentId = studentId, SubjectId = subjectId, Value = (float) value};
            this._gradeRepository.Add(grade);
        }

        [XmlRpcMethod("service.DeleteGrade", Description = "Deletes grade with given id from the database")]
        public void DeleteGrade(int gradeId)
        {
            Console.WriteLine("New DeleteGrade Request " + DateTime.Now);
            this._gradeRepository.Delete(gradeId);
        }

        [XmlRpcMethod("service.UpdateGrade", Description = "Updates grade with given id")]
        public void UpdateGrade(int gradeId, double value)
        {
            Console.WriteLine("New UpdateGrade Request " + DateTime.Now);
            this._gradeRepository.Update(gradeId, (float) value);
        }

        [XmlRpcMethod("service.GetAllGrades", Description = "Retrieves all grades from the database")]
        public XmlRpcStruct[] GetAllGrades()
        {
            Console.WriteLine("New GetAllGrades Request " + DateTime.Now);
            return Converter.GradesToXmlRpcStructs(this._gradeRepository.GetAll());
        }
    }
}